man/readelf-ng.1
.TH READELF-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
readelf-ng: display information about ELF files.

.SH SYNOPSIS
readelf-ng [\fB\-H\fR|\fB\-\-help\fR]
           [\fB\-v\fR|\fB\-\-version\fR]
           [\fB\-A\fR|\fB\-\-arch-specific\fR]
           [\fB\-I\fR|\fB\-\-histogram\fR]
           [\fB\-S\fR|\fB\-\-section-headers\fR]
           [\fB\-V\fR|\fB\-\-version-info\fR]
           [\fB\-a\fR|\fB\-\-all\fR]
           [\fB\-d\fR|\fB\-\-dynamic\fR]
           [\fB\-e\fR|\fB\-\-headers\fR]
           [\fB\-g\fR|\fB\-\-section-groups\fR]
           [\fB\-t\fR|\fB\-\-section-details\fR]
           [\fB\-h\fR|\fB\-\-file-header\fR]
           [\fB\-l\fR|\fB\-\-program-headers\fR]
           [\fB\-n\fR|\fB\-\-notes\fR]
           [\fB\-r\fR|\fB\-\-relocs\fR]
           [\fB\-s\fR|\fB\-\-symbols\fR]
           [\fB\-u\fR|\fB\-\-unwind\fR]
           [\fB\-D\fR|\fB\-\-use-dynamic\fR]
           [\fB\-C\fR|\fB\-\-demangle\fR]
           [\fB\-x\fR <number or name>|\fB\-\-hex\-dump=\fR<number or name>]
           [\fB\-Z\fR <number or name>|\fB\-\-code\-dump=\fR<number or name>]
           [\fB\-p\fR <number or name>|\fB\-\-string\-dump=\fR<number or name>]
           [\fB\-U\fR <number or name>|\fB\-\-unicode\-dump=\fR<number or name>]
           [\fB\-R\fR <number or name>|\fB\-\-relocated\-dump=\fR<number or name>]
           [\fB\-m\fR <number or name>|\fB\-\-disassemble=\fR<number or name>]
           [\fB\-z\fR|\fB\-\-decompress\fR]
           [\fB\-I\fR|\fB\-\-info\fR]
           [\fB\-X\fR|\fB\-\-hash\fR]
           [\fB\-E\fR|\fB\-\-entropy\fR]
           [\fB\-\-verbose\fR]
           [\fB\-w[aAcFikKLmrRsptTuUgX]\fR|
            \fB\-\-debug-dump\fR[=abbrev,=addr,=cu_index,=frames\-interp,=info,=links,=follow\-links,
                         =decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,
                         =trace_abbrev,=trace_info,=gdb_index,=enhanced]]
           [\fB\-M\fR <options>|\fB\-\-disassembler\-options=\fR<options>]
           [\fB\-T\fR <options>|\fB\-\-script=\fR<options>]
           \fIelffile\fR...

.SH DESCRIPTION
\fBreadelf-ng\fR displays information about one or more ELF format object files. The options control what particular information to display.

\fIelffile...\fR are the object files to be examined. 32-bit and 64-bit ELF files are supported, as are archives containing ELF files.

.SH OPTIONS
.IP "\fB\-H\fR" 4
.IX Item "-H"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to readelf-ng and exit.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of readelf-ng and exit.
.IP "\fB\-a\fR" 4
.IX Item "-a"
.PD 0
.IP "\fB\-\-all\fR" 4
.IX Item "--all"
.PD
Equivalent to specifying: \fB\-\-arch-specific\fR, \fB\-\-histogram\fR, \fB\-\-section-headers\fR, \fB\-\-version-info\fR, \fB\-\-dynamic\fR, \fB\-\-section-groups\fR, \fB\-\-file-header\fR, \fB\-\-program-headers\fR, \fB\-\-notes\fR, \fB\-\-relocs\fR, \fB\-\-symbols\fR and \fB\-\-unwind\fR.
.IP "\fB\-A\fR" 4
.IX Item "-A"
.PD 0
.IP "\fB\-\-arch-specific\fR" 4
.IX Item "--arch-specific"
.PD
Displays architecture-specific information in the file.
.IP "\fB\-I\fR" 4
.IX Item "-I"
.PD 0
.IP "\fB\-\-histogram\fR" 4
.IX Item "--histogram"
.PD
Display a histogram of bucket list lengths when displaying the contents of the symbol tables.
.IP "\fB\-S\fR" 4
.IX Item "-S"
.PD 0
.IP "\fB\-\-section\-headers\fR" 4
.IX Item "--section-headers"
.PD
Displays the information contained in the file's section headers.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-version\-info\fR" 4
.IX Item "--version-info"
.PD
Displays the contents of the version sections in the file.
.IP "\fB\-d\fR" 4
.IX Item "-d"
.PD 0
.IP "\fB\-\-dynamic\fR" 4
.IX Item "--dynamic"
.PD
Displays the contents of the file's dynamic section.
.IP "\fB\-e\fR" 4
.IX Item "-e"
.PD 0
.IP "\fB\-\-headers\fR" 4
.IX Item "--headers"
.PD
Display all the headers in the file.
Equivalent to specifying: \fB\-\-section-headers\fR, \fB\-\-file-header\fR and \fB\-\-program-headers\fR.
.IP "\fB\-g\fR" 4
.IX Item "-g"
.PD 0
.IP "\fB\-\-section\-groups\fR" 4
.IX Item "--section-groups"
.PD
Displays the information contained in the file's section groups.
.IP "\fB\-t\fR" 4
.IX Item "-t"
.PD 0
.IP "\fB\-\-section\-details\fR" 4
.IX Item "--section-details"
.PD
Displays the detailed section information. Implies \fB\-S\fR.
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD 0
.IP "\fB\-\-file\-header\fR" 4
.IX Item "--file-header"
.PD
Displays the information contained in the ELF header at the start of the file.
.IP "\fB\-l\fR" 4
.IX Item "-l"
.PD 0
.IP "\fB\-\-program\-headers\fR" 4
.IX Item "--program-headers"
.PD
Displays the information contained in the file's segment headers.
.IP "\fB\-n\fR" 4
.IX Item "-n"
.PD 0
.IP "\fB\-\-notes\fR" 4
.IX Item "--notes"
.PD
Displays the contents of the NOTE segments.
.IP "\fB\-r\fR" 4
.IX Item "-r"
.PD 0
.IP "\fB\-\-relocs\fR" 4
.IX Item "--relocs"
.PD
Displays the contents of the file's relocation section.
.IP "\fB\-s\fR" 4
.IX Item "-s"
.PD 0
.IP "\fB\-\-symbols\fR" 4
.IX Item "--symbols"
.PD
Displays the entries in symbol table section of the file.
.IP "\fB\-u\fR" 4
.IX Item "-u"
.PD 0
.IP "\fB\-\-unwind\fR" 4
.IX Item "--unwind"
.PD
Displays the contents of the file's unwind section.
.IP "\fB\-D\fR" 4
.IX Item "-D"
.PD 0
.IP "\fB\-\-use\-dynamic\fR" 4
.IX Item "--use-dynamic"
.PD
Use the dynamic section info when displaying symbols.
.IP "\fB\-C\fR" 4
.IX Item "-C"
.PD 0
.IP "\fB\-\-demangle\fR" 4
.IX Item "--demangle"
.PD
Decode (demangle) low-level symbol names into user-level names. This makes C++ function names readable. Different compilers have different mangling styles.
.IP "\fB\-x <number or name>\fR" 4
.IX Item "-x <number or name>"
.PD 0
.IP "\fB\-\-hex\-dump=<number or name>\fR" 4
.IX Item "--hex-dump=<number or name>"
.PD
Displays the contents of the indicated section as a hexadecimal bytes.
.IP "\fB\-Z <number or name>\fR" 4
.IX Item "-Z <number or name>"
.PD 0
.IP "\fB\-\-code\-dump=<number or name>\fR" 4
.IX Item "--code-dump=<number or name>"
.PD
Displays the contents of the indicated section as C Language Shell Code.
.IP "\fB\-p <number or name>\fR" 4
.IX Item "-p <number or name>"
.PD 0
.IP "\fB\-\-string\-dump=<number or name>\fR" 4
.IX Item "--string-dump=<number or name>"
.PD
Displays the contents of the indicated section as printable strings.
.IP "\fB\-u <number or name>\fR" 4
.IX Item "-u <number or name>"
.PD 0
.IP "\fB\-\-unicode\-dump=<number or name>\fR" 4
.IX Item "--unicode-dump=<number or name>"
.PD
Displays the contents of the indicated section as printable strings.
.IP "\fB\-R <number or name>\fR" 4
.IX Item "-R <number or name>"
.PD 0
.IP "\fB\-\-relocated\-dump=<number or name>\fR" 4
.IX Item "--relocated-dump=<number or name>"
.PD
Displays the contents of the indicated section as a hexadecimal bytes.
.IP "\fB\-z\fR" 4
.IX Item "-z"
.PD 0
.IP "\fB\-\-decompress\fR" 4
.IX Item "--decompress"
.PD
Requests that the section(s) being dumped by \fBx\fR, \fBR\fR or \fBp\fR options are decompressed before being displayed.
.IP "\fB\-I\fR" 4
.IX Item "-I"
.PD 0
.IP "\fB\-\-info\fR" 4
.IX Item "--info"
.PD
Equivalent to specifying: --hash --entropy
.IP "\fB\-X\fR" 4
.IX Item "-X"
.PD 0
.IP "\fB\-\-hash\fR" 4
.IX Item "--hash"
.PD
Calculates and displays the SHA-256 hash.
.IP "\fB\-E\fR" 4
.IX Item "-E"
.PD 0
.IP "\fB\-\-entropy\fR" 4
.IX Item "--entropy"
.PD
Displays the entropy of the object.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Print additional information.
.IP "\fB\-w[aAcFikKLmrRsptTuUgX]\fR" 4
.IX Item "-w[aAcFikKLmrRsptTuUgX]"
.PD 0
.IP "\fB\-\-debug-dump[=abbrev,=addr,=cu_index,=frames-interp,=info,=links,=follow-links,=decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,=trace_abbrev,=trace_info,=gdb_index,=enhanced]\fR" 4
.IX Item "--debug-dump[=abbrev,=addr,=cu_index,=frames-interp,=info,=links,=follow-links,=decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,=trace_abbrev,=trace_info,=gdb_index,=enhanced]"
.PD
Displays the contents of the \s-1DWARF\s0 debug sections in the file, if any are present.
.RS 4
.ie n .IP """a""" 4
.IX Item "a"
.PD 0
.ie n .IP """=abbrev""" 4
.IX Item "=abbrev"
.PD
Displays the contents of the \fB.debug_abbrev\fR section.
.ie n .IP """A""" 4
.IX Item "A"
.PD 0
.ie n .IP """=addr""" 4
.IX Item "=addr"
.PD
Displays the contents of the \fB.debug_addr\fR section.
.ie n .IP """c""" 4
.IX Item "c"
.PD 0
.ie n .IP """=cu_index""" 4
.IX Item "=cu_index"
.PD
Displays the contents of the \fB.debug_cu_index\fR|\fB.debug_tu_index\fR sections.
.ie n .IP """F""" 4
.IX Item "F"
.PD 0
.ie n .IP """=frames\-interp""" 4
.IX Item "=frames-interp"
.PD
Display the interpreted contents of a \fB.debug_frame\fR section.
.ie n .IP """i""" 4
.IX Item "i"
.PD 0
.ie n .IP """=info""" 4
.IX Item "=info"
.PD
Displays the contents of the \fB.debug_info\fR section.
.ie n .IP """k""" 4
.IX Item "k"
.PD 0
.ie n .IP """=links""" 4
.IX Item "=links"
.PD
Displays the contents of the \fB.gnu_debuglink\fR|\fB.gnu_debugaltlink\fR sections.
.ie n .IP """K""" 4
.IX Item "K"
.PD 0
.ie n .IP """=follow\-links""" 4
.IX Item "=follow-links"
.PD
Display the contents of any selected debug sections that are found in linked, separate debug info file(s).
.ie n .IP """L""" 4
.IX Item "L"
.PD 0
.ie n .IP """=decodedline""" 4
.IX Item "=decodedline"
.PD
Displays the interpreted contents of the \fB.debug_line\fR section.
.ie n .IP """m""" 4
.IX Item "m"
.PD 0
.ie n .IP """=macro""" 4
.IX Item "=macro"
.PD
Displays the contents of the \fB.debug_macro\fR|\fB.debug_macinfo\fR sections.
.ie n .IP """r""" 4
.IX Item "r"
.PD 0
.ie n .IP """=aranges""" 4
.IX Item "=aranges"
.PD
Displays the contents of the \fB.debug_aranges\fR section.
.ie n .IP """R""" 4
.IX Item "R"
.PD 0
.ie n .IP """=ranges""" 4
.IX Item "=ranges"
.PD
Displays the contents of the \fB.debug_ranges\fR|\fB.debug_rnglists\fR sections.
.ie n .IP """s""" 4
.IX Item "s"
.PD 0
.ie n .IP """=str""" 4
.IX Item "=str"
.PD
Displays the contents of the \fB.debug_str\fR|\fB.debug_line_str\fR|\fB.debug_str_offsets\fR sections.
.ie n .IP """p""" 4
.IX Item "p"
.PD 0
.ie n .IP """=pubnames""" 4
.IX Item "=pubnames"
.PD
Displays the contents of the \fB.debug_pubnames\fR|\fB.debug_gnu_pubnames\fR sections.
.ie n .IP """t""" 4
.IX Item "t"
.PD 0
.ie n .IP """=pubtypes""" 4
.IX Item "=pubtypes"
.PD
Displays the contents of the \fB.debug_pubtypes\fR|\fB.debug_gnu_pubtypes\fR sections.
.ie n .IP """T""" 4
.IX Item "T"
.PD 0
.ie n .IP """=trace_aranges""" 4
.IX Item "=trace_aranges"
.PD
Displays the contents of the \fB.trace_aranges\fR section.
.ie n .IP """u""" 4
.IX Item "u"
.PD 0
.ie n .IP """=trace_abbrev""" 4
.IX Item "=trace_abbrev"
.PD
Displays the contents of the \fB.trace_abbrev\fR section.
.ie n .IP """U""" 4
.IX Item "U"
.PD 0
.ie n .IP """=trace_info""" 4
.IX Item "=trace_info"
.PD
Displays the contents of the \fB.trace_info\fR section.
.ie n .IP """g""" 4
.IX Item "g"
.PD 0
.ie n .IP """=gdb_index""" 4
.IX Item "=gdb_index"
.PD
Displays the contents of the \fB.gdb_index\fR|\fB.debug_names\fR sections.
.ie n .IP """X""" 4
.IX Item "X"
.PD 0
.ie n .IP """=enhanced""" 4
.IX Item "=enhanced"
.PD
Displays the contents in the enhanced mode format.
.RE
.IP "\fB\-m <number or name>\fR" 4
.IX Item "-m <number or name>"
.PD 0
.IP "\fB\-\-disassemble=<number or name>\fR" 4
.IX Item "--disassemble=<number or name>"
.PD
Display the assembler mnemonics for the machine instructions from the input file.
.IP "\fB\-M <options>\fR" 4
.IX Item "-M <options>"
.PD 0
.IP "\fB\-\-disassembler\-options=<options>\fR" 4
.IX Item "--disassembler-options=<options>"
.PD
Pass target specific information to the disassembler.
.RS 4
.ie n .IP """att""" 4
.IX Item "att"
.PD
Select the AT&T syntax mode.
.ie n .IP """intel""" 4
.IX Item "intel"
.PD
Select the Intel syntax mode.
.ie n .IP """arm32""" 4
.IX Item "arm32"
.PD 0
.ie n .IP """arm64""" 4
.IX Item "arm64"
.ie n .IP """i386""" 4
.IX Item "i386"
.ie n .IP """i8086""" 4
.IX Item "i8086"
.ie n .IP """x86-64""" 4
.IX Item "x86-64"
.ie n .IP """riscv32""" 4
.IX Item "riscv32"
.ie n .IP """riscv64""" 4
.IX Item "riscv64"
.ie n .IP """mips1""" 4
.IX Item "mips1"
.ie n .IP """mips2""" 4
.IX Item "mips2"
.ie n .IP """mips3""" 4
.IX Item "mips3"
.ie n .IP """mips4""" 4
.IX Item "mips4"
.ie n .IP """mips5""" 4
.IX Item "mips5"
.ie n .IP """mips32""" 4
.IX Item "mips32"
.ie n .IP """mips32r2""" 4
.IX Item "mips32r2"
.ie n .IP """mips32r6""" 4
.IX Item "mips32r6"
.ie n .IP """mips64""" 4
.IX Item "mips64"
.ie n .IP """mips64r2""" 4
.IX Item "mips64r2"
.ie n .IP """mips64r6""" 4
.IX Item "mips64r6"
.PD
Select disassembly for the given architecture.
.ie n .IP """amd64""" 4
.IX Item "amd64"
.PD 0
.ie n .IP """intel64""" 4
.IX Item "intel64"
.PD
Select between AMD64 ISA and Intel64 ISA.
.ie n .IP """addr16""" 4
.IX Item "addr16"
.PD 0
.ie n .IP """addr32""" 4
.IX Item "addr32"
.ie n .IP """addr64""" 4
.IX Item "addr64"
.ie n .IP """data16""" 4
.IX Item "data16"
.ie n .IP """data32""" 4
.IX Item "data32"
.PD
Specify the default address size and operand size.
.RE
.IP "\fB\-T <options>\fR" 4
.IX Item "-T <options>"
.PD 0
.IP "\fB\-\-script=<options>\fR" 4
.IX Item "--script=<options>"
.PD
Generate simple scripts to modify any section to reverse obfuscation.
.RS 4
.ie n .IP """add8""" 4
.IX Item "add8"
.PD 0
.ie n .IP """add16""" 4
.IX Item "add16"
.ie n .IP """add32""" 4
.IX Item "add32"
.PD
ADD (addition) the input with the given key.
.ie n .IP """not8""" 4
.IX Item "not8"
.PD 0
.ie n .IP """not16""" 4
.IX Item "not16"
.ie n .IP """not32""" 4
.IX Item "not32"
.PD
NOT (negation) the input.
.ie n .IP """rol8""" 4
.IX Item "rol8"
.PD 0
.ie n .IP """rol16""" 4
.IX Item "rol16"
.ie n .IP """rol32""" 4
.IX Item "rol32"
.PD
ROL (rotate-left) the input with the given key.
.ie n .IP """ror8""" 4
.IX Item "ror8"
.PD 0
.ie n .IP """ror16""" 4
.IX Item "ror16"
.ie n .IP """ror32""" 4
.IX Item "ror32"
.PD
ROR (rotate-right) the input with the given key.
.ie n .IP """rot5""" 4
.IX Item "rot5"
.PD 0
.ie n .IP """rot13""" 4
.IX Item "rot13"
.ie n .IP """rot18""" 4
.IX Item "rot18"
.ie n .IP """rot47""" 4
.IX Item "rot47"
.PD
Caesar substitution cipher which rotates alphabet characters by the specified amount.
.ie n .IP """shl8""" 4
.IX Item "shl8"
.PD 0
.ie n .IP """shl16""" 4
.IX Item "shl16"
.ie n .IP """shl32""" 4
.IX Item "shl32"
.PD
SHL (left-shift) the input with the given key.
.ie n .IP """shr8""" 4
.IX Item "shr8"
.PD 0
.ie n .IP """shr16""" 4
.IX Item "shr16"
.ie n .IP """shr32""" 4
.IX Item "shr32"
.PD
SHR (right-shift) the input with the given key.
.ie n .IP """sub8""" 4
.IX Item "sub8"
.PD 0
.ie n .IP """sub16""" 4
.IX Item "sub16"
.ie n .IP """sub32""" 4
.IX Item "sub32"
.PD
SUB (subtract) the input with the given key.
.ie n .IP """xor8""" 4
.IX Item "xor8"
.PD 0
.ie n .IP """xor16""" 4
.IX Item "xor16"
.ie n .IP """xor32""" 4
.IX Item "xor32"
.PD
XOR (eXclusive OR) the input with the given key.
.ie n .IP """dec""" 4
.IX Item "dec"
.PD 0
.ie n .IP """inc""" 4
.IX Item "inc"
.PD
Amount to increment or decrement the given key.

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/readelf-ng.c
#include "buffer.h"
#include "printf.h"
#include "readar.h"
#include "scripts.h"
#include "readelf.h"
#include "options.h"

#include "static/readelf-ng.ci"

#define THIS_NAME "readelf-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zDESCRIPTIONELF);
  n += usage_synopsis0(o, name, args);
  n += usage_synopsis1(o, name, zDEBUGELFARGS, zREADELFARGS0, zREADELFARGS1);
  n += usage_synopsis2(o, name, zDISASSEMBLEARGS, zREADELFARGS2, zREADELFARGS3);
  n += usage_synopsis2(o, name, zSCRIPTCOMMANDS, zREADELFARGS4, zREADELFARGS5);
  n += usage_objname(o, name, "elffile");
  n += printf_eol();
  n += usage_description(o, name, args, zELFDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_options1(o, name, zDEBUGELFARGS, zREADELFARGS0, zREADELFARGS1);
  n += usage_options2(o, name, zDISASSEMBLEARGS, zREADELFARGS2, zREADELFARGS3);
  n += usage_options2(o, name, zSCRIPTCOMMANDS, zREADELFARGS4, zREADELFARGS5);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_readelf(poptions_t o, int argc, char** argv, char* name) {
  if (0 == argc) {
    return usage(o, THIS_NAME, zREADELFARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(o->prgname);

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      MALLOCA(char, arg0, 1024);
      MALLOCA(char, arg1, 1024);

      if (ECODE_ISOK(breakup_args(argv[i], arg0, NELEMENTS(arg0), arg1, NELEMENTS(arg1)))) {
        if (0 == xstrcmp(arg0, zREADELFARGS1)) {
          imode_t ocdump = get_options2(o, zDEBUGELFARGS, arg1);
          if (0 == ocdump) {
            return odeath(o, THIS_NAME, arg1);
          }
          o->ocdump |= ocdump;
        } else if (0 == xstrcmp(arg0, zREADELFARGS3)) {
          imode_t ocdump = get_options2(o, zDISASSEMBLEARGS, arg1);
          if (0 == ocdump) {
            return odeath(o, THIS_NAME, arg1);
          }
          o->ocdump |= ocdump;
        } else if (0 == xstrcmp(arg0, zREADELFARGS5)) {
          sinsert(o, arg1, zSCRIPTCOMMANDS, NULL);
        } else if (0 == xstrcmp(arg0, "--hex-dump")) {
          oinsertsname(o, ACT_HEXDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--string-dump")) {
          oinsertsname(o, ACT_STRDUMP8, arg1);
        } else if (0 == xstrcmp(arg0, "--unicode-dump")) {
          oinsertsname(o, ACT_STRDUMP16, arg1);
        } else if (0 == xstrcmp(arg0, "--relocated-dump")) {
          oinsertsname(o, ACT_RELDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--code-dump")) {
          oinsertsname(o, ACT_CODEDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--disassemble")) {
          oinsertsname(o, ACT_DISASSEMBLE, arg1);
        } else if (0 == xstrcmp(arg0, "--decompress")) {
          oinsertsname(o, ACT_ZLIB, arg1);
        } else {
          return odeath(o, THIS_NAME, arg0 + 2);
        }
      } else {
        imode_t action = get_options2(o, zREADELFARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 2);
        }
        o->action |= action;
      }
    } else if ('-' == argv[i][0]) {
      if (argv[i][0] == zREADELFARGS0[0] && argv[i][1] == zREADELFARGS0[1]) {
        imode_t ocdump = get_options1(o, zDEBUGELFARGS, argv[i] + 1);
        if (0 == ocdump && argv[i][2]) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->ocdump |= ocdump ? ocdump : set_options1(o, zDEBUGELFARGS);
      } else if ( argv[i][0] == zREADELFARGS2[0] && argv[i][1] == zREADELFARGS2[1]  ) {
        imode_t ocdump = get_options2(o, zDISASSEMBLEARGS, argv[++i]);
        if (0 == ocdump) {
          return odeath(o, THIS_NAME, argv[i]);
        }
        o->ocdump |= ocdump;
      } else if (argv[i][0] == zREADELFARGS4[0] && argv[i][1] == zREADELFARGS4[1]) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        sinsert(o, argv[++i], zSCRIPTCOMMANDS, NULL);
      } else if (0 == xstrcmp(argv[i], "-x")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_HEXDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-p")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_STRDUMP8, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-U")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_STRDUMP16, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-R")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_RELDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-Z")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_CODEDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-m")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_DISASSEMBLE, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-z")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_ZLIB, argv[++i]);
      } else {
        imode_t action = get_options1(o, zREADELFARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->action |= action;
      }
    } else if (0 == o->inpname[0]) {
      xstrncpy(o->inpname, argv[i], NELEMENTS(o->inpname));
    } else {
      return odeath(o, THIS_NAME, argv[i]);
    }
  }

  o->ocdump |= get_ocdump(o, zREADELFSWAP, o->action);

  if (o->action & OPTPROGRAM_VERSION) {
    return version0(o, THIS_NAME, zREADELFARGS);
  }

  if (o->action & OPTPROGRAM_HELP) {
    return usage(o, THIS_NAME, zREADELFARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_readelf(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname[0]) {
      pbuffer_t p = bopen(o->inpname);
      if (p) {
        if (isAR(p)) {
          x = readar(p, o);
        } else if (isELF(p)) {
          x = readelf(p, o);
        } else {
          printf_e("'%s': invalid file format.", o->inpname);
        }
      } else {
        printf_e("'%s': no such file.", o->inpname);
      }

     bfree(p);
    }

    ofree(o);
  }

  return x;
}

